servers/rendering/renderer_rd/effects/metal_fx.mm

```diff
@@ -0,0 +1,214 @@
+}
+
+MFXSpatialEffect::~MFXSpatialEffect() {
+}
+
+void MFXSpatialEffect::callback(RDD *p_driver, RDD::CommandBufferID p_command_buffer, CallbackArgs *p_userdata) {
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wunguarded-availability"
+
+	MDCommandBuffer *obj = (MDCommandBuffer *)(p_command_buffer.id);
+	obj->end();
+
+	id<MTLTexture> src_texture = rid::get(p_userdata->src);
+	id<MTLTexture> dst_texture = rid::get(p_userdata->dst);
+
+	__block id<MTLFXSpatialScaler> scaler = p_userdata->ctx.scaler;
+	scaler.colorTexture = src_texture;
+	scaler.outputTexture = dst_texture;
+	[scaler encodeToCommandBuffer:obj->get_command_buffer()];
+	// TODO(sgc): add API to retain objects until the command buffer completes
+	[obj->get_command_buffer() addCompletedHandler:^(id<MTLCommandBuffer> _Nonnull) {
+		// This block retains a reference to the scaler until the command buffer.
+		// completes.
+		scaler = nil;
+	}];
+
+#pragma clang diagnostic pop
+}
+
+void MFXSpatialEffect::process(MFXSpatialContext *p_ctx, RID p_src, RID p_dst) {
+	CallbackArgs *userdata = &args[i];
+	userdata->src.id = RD::get_singleton()->get_driver_resource(RDC::DRIVER_RESOURCE_TEXTURE, p_src);
+	userdata->dst.id = RD::get_singleton()->get_driver_resource(RDC::DRIVER_RESOURCE_TEXTURE, p_dst);
+	userdata->ctx = *p_ctx;
+	i = (i + 1) % RID_COUNT;
+	RD::CallbackResource res[2] = {
+		{ .rid = p_src, .usage = RD::CALLBACK_RESOURCE_USAGE_TEXTURE_SAMPLE },
+		{ .rid = p_dst, .usage = RD::CALLBACK_RESOURCE_USAGE_STORAGE_IMAGE_READ_WRITE }
+	};
+	RD::get_singleton()->driver_callback_add((RDD::DriverCallback)MFXSpatialEffect::callback, userdata, VectorView<RD::CallbackResource>(res, 2));
+}
+
+MFXSpatialContext *MFXSpatialEffect::create_context(CreateParams p_params) const {
+	DEV_ASSERT(RD::get_singleton()->has_feature(RD::SUPPORTS_METALFX_SPATIAL));
+
+	if (@available(macOS 13.0, iOS 16.0, *)) {
+		RenderingDeviceDriverMetal *rdd = (RenderingDeviceDriverMetal *)RD::get_singleton()->get_device_driver();
+		PixelFormats &pf = rdd->get_pixel_formats();
+		id<MTLDevice> dev = rdd->get_device();
+
+		MTLFXSpatialScalerDescriptor *desc = [MTLFXSpatialScalerDescriptor new];
+		desc.inputWidth = (NSUInteger)p_params.input_size.width;
+		desc.inputHeight = (NSUInteger)p_params.input_size.height;
+
+		desc.outputWidth = (NSUInteger)p_params.output_size.width;
+		desc.outputHeight = (NSUInteger)p_params.output_size.height;
+
+		desc.colorTextureFormat = pf.getMTLPixelFormat(p_params.input_format);
+		desc.outputTextureFormat = pf.getMTLPixelFormat(p_params.output_format);
+		desc.colorProcessingMode = MTLFXSpatialScalerColorProcessingModeLinear;
+		id<MTLFXSpatialScaler> scaler = [desc newSpatialScalerWithDevice:dev];
+		MFXSpatialContext *context = memnew(MFXSpatialContext);
+		context->scaler = scaler;
+		return context;
+	} else {
+		// Shouldn't be possible to get to this point.
+		CRASH_NOW_MSG("MetalFX spatial scaler are only available on macOS 13.0, iOS 16.0 or later.");
```

Comment:
Is it possible some compiler could complain about this control path not returning anything even if it's supposed to crash?

Reply:
This will only get compiled on macOS, so it will be ok. I included it as a guard to indicate a programming error

Reply:
The intention behind it is clear, I was wondering if we're guaranteed the compiler will never complain about it. Something like this does result in an error due to being a warning in other platforms. I think for the sake of guarding against a possible future build error or against other types of analysis we should be returning in all control paths anyway.

Reply:
Ahh, I'll just remove the platform guard, as we already have the DEV_ASSERT 👍🏻

Reply:
Done!